tm2/pkg/sdk/auth/genesis.go

```diff
@@ -0,0 +1,58 @@
+package auth
+
+import (
+	"github.com/gnolang/gno/tm2/pkg/amino"
+	"github.com/gnolang/gno/tm2/pkg/sdk"
+)
+
+// GenesisState - all state that must be provided at genesis
+type GenesisState struct {
+	Params Params `json:"params" yaml:"params"`
+}
+
+// NewGenesisState - Create a new genesis state
+func NewGenesisState(params Params) GenesisState {
+	return GenesisState{params}
+}
+
+// DefaultGenesisState - Return a default genesis state
+func DefaultGenesisState() GenesisState {
+	return NewGenesisState(DefaultParams())
+}
+
+// ValidateGenesis performs basic validation of genesis data returning an
+// error for any failed validation criteria.
+func ValidateGenesis(data GenesisState) error {
+	return data.Params.Validate()
+}
+
+// InitGenesis - Init store state from genesis data
+func (ak AccountKeeper) InitGenesis(ctx sdk.Context, data GenesisState) {
+	if amino.DeepEqual(data, GenesisState{}) {
+		return
+	}
+
+	if err := ValidateGenesis(data); err != nil {
+		panic(err)
+	}
+
+	// The unrestricted address must have been created as one of the genesis accounts.
+	// Otherwise, we cannot verify the unrestricted address in the genesis state.
+
+	for _, addr := range data.Params.UnrestrictedAddrs {
+		acc := ak.GetAccount(ctx, addr)
```

Comment:
There is tight coupling here. The account keeper must have a previously initialized account in order for us to modify it in another func

Can't we keep the store of unrestricted accounts somewhere, instead of keeping the `Unrestricted` information on the account itself?

My biggest concern is that we're coupling temporary logic (account balance transfer locks) into a structure that will be encoded and saved permanently to disk. Every time we use the account, even in 10 years, we will have to keep track of its "restricted state".

This is why I'm suggesting we go with an approach that isn't so coupled

Reply:
There are two reasons for implementing it this way:

1) There was a previous requirement to track and unlock token transfers on an individual account basis, not just for whitelisted accounts. That is the reason the restricted state is on the account level.  We can revisit the requirement, otherwise, we will need to track the restricted state for a long time. 

2) The whitelisted unrestricted accounts need to be validated to ensure they exist when we load the genesis state.